driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.async;
+
+import com.mongodb.internal.async.function.RetryState;
+import com.mongodb.internal.async.function.RetryingAsyncCallbackSupplier;
+
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * See AsyncRunnableTest for usage
+ */
+public interface AsyncRunnable {
```

Comment:
There is the `com.mongodb.internal.async.function` package that contains asynchronous callback-based counterparts for the following Java SE synchronous functional interfaces (these async counterparts are also functional interfaces):

- `Function` - `AsyncCallbackFunction`
- `BiFunction` - `AsyncCallbackBiFunction`
- `Runnable` - `AsyncCallbackRunnable`
- `Supplier` - `AsyncCallbackSupplier`

as well as APIs for somewhat emulating (not necessary with exactly the same semantics) the following Java language constructs in the asynchronous callback-based code:

- `while (true)` - `AsyncCallbackLoop`
- `while (true)` with `try-catch` in the loop body - `RetryingAsyncCallbackSupplier`
- `try-finally` - `AsyncCallbackRunnable.whenComplete`/`AsyncCallbackSupplier.whenComplete`


`AsyncRunnable`, `AsyncSupplier` introduced in this PR are also functional interfaces with the virtual methods being identical (barring names, which are inconsistent with the Java SE API for the interfaces introduced in this PR) to those of `AsyncCallbackRunnable`, `AsyncCallbackSupplier`. The concrete (`default`) methods on the `AsyncRunnable`, `AsyncSupplier` interfaces are meant to simplify writing asynchronous callback-based code, similarly to what the concrete methods on `AsyncCallbackRunnable`, `AsyncCallbackSupplier` are meant for. In fact, `AsyncRunnable.completeAlways` is similar to `AsyncCallbackRunnable.whenComplete`, but with issues described in a different thread. These methods could have been introduced on the existing interfaces in the existing package.

We talked about this a month ago, and at that time I failed to understand the reason for introducing `AsyncRunnable`, `AsyncSupplier` instead of using / developing further the existing interfaces. I still don't understand the reason, could you please explain it here?

Reply:
When merging, can we preserve the method names similar to those in sync counterparts (`Function.apply`, `Supplier.get`, `Runnable.run`, etc.) for unchecked methods (so `uncheckedApply`, `uncheckedGet`, `uncheckedRun`, etc.)?